src/appshell/internal/startupscenario.cpp

```diff
@@ -94,50 +95,48 @@ muse::async::Promise<muse::Ret> StartupScenario::runOnSplashScreen()
         registerAudioPlugins();
 
         if (multiInstancesProvider()->instances().size() != 1) {
-            const Ret ret = muse::make_ret(Ret::Code::Cancel);
+            const Ret ret = muse::make_ret(Ret::Code::Ok);
             return resolve(ret);
         }
 
         // Calculate the total number of expected update checks (TODO: check the
         // connection before trying any of this)...
-        static size_t totalChecksExpected = 0;
 
         const bool canCheckAppUpdate = appUpdateScenario() && appUpdateScenario()->needCheckForUpdate();
-        if (canCheckAppUpdate) {
-            ++totalChecksExpected;
-        }
-
         const bool canCheckMuseSoundsUpdate = museSoundsUpdateScenario() && museSoundsUpdateScenario()->needCheckForUpdate();
-        if (canCheckMuseSoundsUpdate) {
-            ++totalChecksExpected;
-        }
-
         //! NOTE: A MuseSampler update check also exists but we run it later (see onStartupPageOpened)...
+        m_totalChecksExpected = size_t(canCheckAppUpdate) + size_t(canCheckMuseSoundsUpdate);
 
-        if (totalChecksExpected == 0) {
+        if (m_totalChecksExpected == 0) {
             const Ret ret = muse::make_ret(Ret::Code::Ok);
             return resolve(ret);
         }
 
+        m_totalChecksReceived = 0;
+
         // Resolve once all checks are completed...
         const auto onUpdateCheckCompleted = [this, resolve](){
-            static size_t totalChecksReceived = 0;
-            IF_ASSERT_FAILED(m_updateCheckInProgress && totalChecksReceived < totalChecksExpected) {
-                m_updateCheckInProgress = false;
-                return;
+            std::unique_lock<std::mutex> lock(m_mutex);
```

Comment:
If I'm not mistaken, all this code runs on the main thread, because `onResolved` is called on the main thread (and the async system calls handlers on the thread from which they were registered). That means that, although we don't know in which order the handlers will run, we do know that only one will run at a time, so no lock is necessary.